lib/Backend/arm64/LegalizeMD.cpp

```diff
@@ -985,6 +985,37 @@ bool LegalizeMD::LegalizeAdrOffset(IR::Instr *instr, uintptr_t instrOffset)
     return true;
 }
 
+bool LegalizeMD::LegalizeDataAdr(IR::Instr *instr, uintptr_t dataOffset)
+{
+    Assert(instr->m_opcode == Js::OpCode::ADR);
+
+    IR::LabelOpnd* labelOpnd = instr->GetSrc1()->AsLabelOpnd();
+    IR::LabelInstr* label = labelOpnd->GetLabel();
+
+    Assert(label->m_isDataLabel);
+
+
+    // dataOffset provides an upper bound on the distance between instr and the label.
+    if (IS_CONST_INT19(dataOffset >> 2))
+    {
+        return false;
+    }
+
+    // The distance is too large to encode as an ADR isntruction so it must be handled as a 3 instruction load immediate. 
+    // The label address won't be know until after encoding. Assign the label opnd as src1 to let the encoder know to handle them as relocs.
+
+    IR::Instr* bits0_15 = IR::Instr::New(Js::OpCode::MOVZ, instr->GetDst(), labelOpnd, IR::IntConstOpnd::New(0, IRType::TyUint8, instr->m_func, true), instr->m_func);
```

Comment:
Note that ADR is sign-extended, while MOVZ will just zero out the high bits; I don't think that this will hit in any cases where we currently use data labels (since they're all forward references?), but it could bite us if we re-arrange the layout of things in the future.

Reply:
This loads the absolute address, not a relative one, so sign extension is not an issue.